File: python/nexus_connector_sdk/sidecar.py
"""Sandboxed legacy Connector Sidecar adapter (SPEC-022 behavior 5).

The sidecar wraps legacy sources - REST, SOAP, GraphQL, SQL read
replicas, ODBC/JDBC, CLI, files, email, webhooks, browser, or desktop
GUI as a last resort - inside a sandbox. The adapter is the port: a
transport implementation talks to a real sidecar process over a
canonical HTTP transport. The sidecar never holds direct authority;
commands stay idempotent and events stay versioned.
"""

from __future__ import annotations

import json
import urllib.error
import urllib.request
from dataclasses import dataclass, field
from typing import Any

from .error import (
    EXTERNAL_PROVIDER,
    TIMEOUT,
    UNAVAILABLE,
    VALIDATION,
    SdkError,
)
from .vocabulary import SidecarTransport
from .wire import InvocationContext

# Canonical sidecar transport protocol version (SPEC-022 Tier 1 REST).
SIDECAR_PROTOCOL_VERSION = "1"

DEFAULT_REQUEST_TIMEOUT_SECONDS = 5.0
DEFAULT_MAX_REQUEST_BYTES = 64 * 1024


@dataclass
class SidecarRequest:
    """A normalized request to a sandboxed sidecar transport."""

    capability_id: str
    transport: str
    action: str
    input: dict[str, Any] = field(default_factory=dict)
    idempotency_key: str | None = None
    context: InvocationContext = field(
        default_factory=lambda: InvocationContext(
            request_id="018f0f6f-9c1e-7b6e-8000-000000000001",
            correlation_id="018f0f6f-9c1e-7b6e-8000-000000000002",
        )
    )

    def to_dict(self) -> dict[str, Any]:
        data: dict[str, Any] = {
            "capability_id": self.capability_id,
            "transport": self.transport,
            "action": self.action,
            "input": self.input,
            "context": self.context.to_dict(),
        }
        # Option field serializes as JSON null (Rust serde parity).
        data["idempotency_key"] = self.idempotency_key
        return data

    @classmethod
    def from_dict(cls, data: dict[str, Any]) -> SidecarRequest:
        return cls(
            capability_id=str(data["capability_id"]),
            transport=str(data["transport"]),
            action=str(data["action"]),
            input=dict(data.get("input", {})),
            idempotency_key=data.get("idempotency_key"),
            context=InvocationContext.from_dict(data["context"]),
        )


@dataclass
class SidecarResponse:
    """Normalized sidecar response."""

    capability_id: str
    output: dict[str, Any] = field(default_factory=dict)
    cursor: str | None = None

    def to_dict(self) -> dict[str, Any]:
        data: dict[str, Any] = {
            "capability_id": self.capability_id,
            "output": self.output,
        }
        # Option field serializes as JSON null (Rust serde parity).
        data["cursor"] = self.cursor
        return data

    @classmethod
    def from_dict(cls, data: dict[str, Any]) -> SidecarResponse:
        return cls(
            capability_id=str(data["capability_id"]),
            output=dict(data.get("output", {})),
            cursor=data.get("cursor"),
        )


class SidecarAdapter:
    """HTTP transport adapter for a sandboxed Connector Sidecar.

    Talks to a real sidecar process over HTTP on a localhost endpoint
    (bind 127.0.0.1, ephemeral port). The transport is versioned
    (``X-Nexus-Protocol-Version``); an unsupported version fails
    closed. Failures are typed and never converted into success.
    """

    def __init__(
        self,
        base_url: str,
        transport: str = SidecarTransport.Rest,
        timeout_seconds: float = DEFAULT_REQUEST_TIMEOUT_SECONDS,
    ) -> None:
        self._base_url = base_url.rstrip("/")
        self._transport = transport
        self._timeout = timeout_seconds

    def transport(self) -> str:
        return self._transport

    def execute(self, request: SidecarRequest) -> SidecarResponse:
        """Execute one normalized sidecar request over real HTTP."""
        body = json.dumps(request.to_dict()).encode("utf-8")
        if len(body) > DEFAULT_MAX_REQUEST_BYTES:
            raise SdkError(
                VALIDATION,
                "sidecar request exceeds bounded size",
                request.context.correlation_id,
                request.context.external_actor_id,
                request.context.tenant_id,
                request.capability_id,
            )
        headers = {
            "Content-Type": "application/json",
            "X-Nexus-Protocol-Version": SIDECAR_PROTOCOL_VERSION,
            "X-Nexus-Correlation-Id": request.context.correlation_id,
        }
        req = urllib.request.Request(
            f"{self._base_url}/v1/execute", data=body, headers=headers, method="POST"
        )
        try:
            with urllib.request.urlopen(req, timeout=self._timeout) as resp:
                payload = json.loads(resp.read().decode("utf-8"))
        except urllib.error.HTTPError as exc:
            raise self._map_http_error(exc, request) from exc
        except urllib.error.URLError as exc:
            raise SdkError(
                UNAVAILABLE,
                f"sidecar unavailable: {exc.reason}",
                request.context.correlation_id,
                request.context.external_actor_id,
                request.context.tenant_id,
                request.capability_id,
            ) from exc
        except TimeoutError as exc:
            raise SdkError(
                TIMEOUT,
                "sidecar request timed out",
                request.context.correlation_id,
                request.context.external_actor_id,
                request.context.tenant_id,
                request.capability_id,
            ) from exc
        return SidecarResponse.from_dict(payload)

    def _map_http_error(self, exc: urllib.error.HTTPError, request: SidecarRequest) -> SdkError:
        code = EXTERNAL_PROVIDER
        try:
            payload = json.loads(exc.read().decode("utf-8"))
            if isinstance(payload, dict) and "code" in payload:
                code = str(payload["code"])
                message = str(payload.get("message", "sidecar error"))
                return SdkError(
                    code,
                    message,
                    payload.get("correlation_id") or request.context.correlation_id,
                    payload.get("actor") or request.context.external_actor_id,
                    payload.get("tenant") or request.context.tenant_id,
                    payload.get("resource") or request.capability_id,
                )
        except json.JSONDecodeError, OSError:
            pass
        return SdkError(
            code,
            f"sidecar HTTP {exc.code}",
            request.context.correlation_id,
            request.context.external_actor_id,
            request.context.tenant_id,
            request.capability_id,
        )


__all__ = [
    "DEFAULT_MAX_REQUEST_BYTES",
    "DEFAULT_REQUEST_TIMEOUT_SECONDS",
    "SIDECAR_PROTOCOL_VERSION",
    "SidecarAdapter",
    "SidecarRequest",
    "SidecarResponse",
]

File: crates/nexus-connector-sdk/src/sidecar.rs
//! Sandboxed legacy Connector Sidecar adapter (SPEC-022 behavior 5).
//!
//! The sidecar wraps legacy sources - REST, SOAP, GraphQL, SQL read
//! replicas, ODBC/JDBC, CLI, files, email, webhooks, browser, or
//! desktop GUI as a last resort - inside a sandbox. The adapter is
//! the port: the transport implementation and sandbox boundary are
//! proven in later EP-011 milestones. The sidecar never holds direct
//! authority; commands stay idempotent and events stay versioned.

use serde::{Deserialize, Serialize};

use nexus_capabilities::context::InvocationContext;

use crate::error::{SdkError, SdkErrorCode};
use crate::vocabulary::SidecarTransport;

/// A normalized request to a sandboxed sidecar transport.
#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
pub struct SidecarRequest {
    /// Capability/connector the request targets.
    pub capability_id: String,
    /// Transport family to wrap.
    pub transport: SidecarTransport,
    /// Canonical action name (never a raw provider method string).
    pub action: String,
    /// Schema-validated canonical input.
    pub input: serde_json::Value,
    /// Idempotency key for commands (SPEC-022 behavior 2).
    pub idempotency_key: Option<String>,
    /// Invocation context.
    pub context: InvocationContext,
}

/// Normalized sidecar response.
#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
pub struct SidecarResponse {
    /// Capability/connector the response belongs to.
    pub capability_id: String,
    /// Schema-validated canonical output.
    pub output: serde_json::Value,
    /// Stable cursor for change/state tracking, when applicable.
    pub cursor: Option<String>,
}

/// Error produced by a sidecar adapter.
#[derive(Debug, Clone, PartialEq, Eq)]
pub struct SidecarAdapterError(pub SdkError);

impl std::fmt::Display for SidecarAdapterError {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        write!(f, "sidecar adapter: {}", self.0)
    }
}

impl std::error::Error for SidecarAdapterError {}

/// Port for a sandboxed legacy connector sidecar transport.
///
/// Implementations wrap exactly one transport family and must
/// normalize provider payloads at the infrastructure boundary: free
/// form provider responses never become domain contracts. Failures
/// are typed and fail closed.
pub trait SidecarAdapter: Send + Sync {
    /// Transport family wrapped by this adapter.
    fn transport(&self) -> SidecarTransport;

    /// Execute one normalized sidecar request.
    fn execute(&self, request: SidecarRequest) -> Result<SidecarResponse, SidecarAdapterError>;
}

/// Construct a typed sidecar error from a raw failure.
pub fn sidecar_error(
    code: SdkErrorCode,
    message: impl Into<String>,
    request: &SidecarRequest,
) -> SidecarAdapterError {
    SidecarAdapterError(SdkError::new(
        code,
        message,
        Some(request.context.correlation_id.to_string()),
        Some(request.context.external_actor_id.clone()),
        Some(request.context.tenant_id.to_string()),
        Some(request.capability_id.clone()),
    ))
}

#[cfg(test)]
mod tests {
    use super::*;
    use nexus_domain::{CorrelationId, NexusId, PrincipalType, TenantId};

    fn ctx() -> InvocationContext {
        InvocationContext::new(
            NexusId::new("018f0f6f-9c1e-7b6e-8000-000000000001").unwrap(),
            CorrelationId::new("018f0f6f-9c1e-7b6e-8000-000000000002").unwrap(),
            None,
            "test",
            "user:alice",
            PrincipalType::Human,
            TenantId::new("018f0f6f-9c1e-7b6e-8000-000000000003").unwrap(),
            Some("mcp".to_string()),
            None,
            None,
            None,
        )
        .unwrap()
    }

    fn req() -> SidecarRequest {
        SidecarRequest {
            capability_id: "legacy.erp".to_string(),
            transport: SidecarTransport::Soap,
            action: "read.invoice".to_string(),
            input: serde_json::json!({ "invoice_id": "INV-1" }),
            idempotency_key: Some("op-1".to_string()),
            context: ctx(),
        }
    }

    struct StubSoap;

    impl SidecarAdapter for StubSoap {
        fn transport(&self) -> SidecarTransport {
            SidecarTransport::Soap
        }
        fn execute(&self, request: SidecarRequest) -> Result<SidecarResponse, SidecarAdapterError> {
            if request.input["invoice_id"] == "INV-1" {
                Ok(SidecarResponse {
                    capability_id: request.capability_id,
                    output: serde_json::json!({ "total": 100 }),
                    cursor: None,
                })
            } else {
                Err(sidecar_error(
                    SdkErrorCode::ExternalProvider,
                    "soap fault",
                    &request,
                ))
            }
        }
    }

    #[test]
    fn ep011_unit_sidecar_adapter_transport_and_execute() {
        let adapter = StubSoap;
        assert_eq!(adapter.transport(), SidecarTransport::Soap);
        let response = adapter.execute(req()).unwrap();
        assert_eq!(response.output["total"], 100);
    }

    #[test]
    fn ep011_unit_sidecar_adapter_fails_closed_typed() {
        let adapter = StubSoap;
        let mut request = req();
        request.input = serde_json::json!({ "invoice_id": "INV-2" });
        let err = adapter.execute(request).unwrap_err();
        assert_eq!(err.0.code, SdkErrorCode::ExternalProvider);
        assert_eq!(err.0.resource.as_deref(), Some("legacy.erp"));
    }

    #[test]
    fn ep011_unit_sidecar_request_serializes_transport() {
        let json = serde_json::to_value(req()).unwrap();
        assert_eq!(json["transport"], "SOAP");
        assert_eq!(json["action"], "read.invoice");
        assert_eq!(json["idempotency_key"], "op-1");
    }
}
